File: src/BackOffice/server/src/services/shared/domain/AggregateRoot.ts
import { DomainEvent } from './DomainEvent';

export interface IAggregateRoot {
  DomainEvents(): Array<DomainEvent>;

  AddDomainEvent(event: DomainEvent): void;

  RemoveDomainEvent(event: DomainEvent): void;

  ClearDomainEvents(): void;

  toPrimitives(): unknown;
}

export abstract class AggregateRoot implements IAggregateRoot {
  private domainEvents: Array<DomainEvent>;

  constructor() {
    this.domainEvents = [];
  }

  DomainEvents = (): Array<DomainEvent> => this.domainEvents;

  AddDomainEvent = (event: DomainEvent): void => {
    this.domainEvents.push(event);
  };

  RemoveDomainEvent = (event: DomainEvent): void => {
    const index: number = this.domainEvents.indexOf(event);

    if (index === -1) return;

    this.domainEvents.splice(this.domainEvents.indexOf(event), 1);
  };

  ClearDomainEvents = () => {
    this.domainEvents = [];
  };

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  abstract toPrimitives(): any;
}

File: src/Services/Shared/MusicStore.Services.Shared/Domain/AggregateRoot.cs
using NetMusicStore.Services.Shared.Interfaces;
using NetMusicStore.Shared.Domain.Bus.Event;

namespace NetMusicStore.Services.Shared { 
    public abstract class AggregateRoot :  IAggregateRoot
    {
        private List<DomainEvent>? _domainEvents;

        public IReadOnlyCollection<DomainEvent>? DomainEvents => _domainEvents?.AsReadOnly();

        public List<DomainEvent> PullDomainEvents()
        {
            var events = _domainEvents;

            _domainEvents = new List<DomainEvent>();

            return events ?? _domainEvents;
        }

        protected void AddDomainEvent(DomainEvent domainEvent)
        {
            _domainEvents?.Add(domainEvent);
        }

        protected void RemoveDomainEvent(DomainEvent domainEvent)
        {
            _domainEvents?.Remove(domainEvent);
        }

        public void ClearDomainEvents()
        {
            _domainEvents?.Clear();
        }

    }
}
